cmd/gh-slack/cmd/api.go

```diff
@@ -0,0 +1,95 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"strings"
+
+	"github.com/cli/go-gh/pkg/config"
+	"github.com/rneatherway/gh-slack/internal/slackclient"
+	"github.com/spf13/cobra"
+)
+
+var apiCmd = &cobra.Command{
+	Use:   "api verb path",
+	Short: "Send an API call to slack",
+	Long:  "Send an API call to slack",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cfg, err := config.Read()
+		if err != nil {
+			return err
+		}
+
+		team, err := getFlagOrElseConfig(cfg, cmd.Flags(), "team")
+		if err != nil {
+			return err
+		}
+
+		if len(args) != 2 {
+			return fmt.Errorf("Expected 2 arguments: verb and path, see help")
+		}
+
+		verb := strings.ToUpper(args[0])
+		path := args[1]
+
+		fields, err := cmd.Flags().GetStringSlice("field")
+		if err != nil {
+			return err
+		}
+
+		mappedFields, err := mapFields(fields)
+		if err != nil {
+			return err
+		}
+
+		logger := log.New(io.Discard, "", log.LstdFlags)
+		if verbose {
+			logger = log.Default()
+		}
+
+		client, err := slackclient.New(team, logger)
+		if err != nil {
+			return err
+		}
+
+		response, err := client.API(verb, path, mappedFields, body)
+		if err != nil {
+			return err
+		}
+
+		fmt.Println(string(response))
+		return nil
+	},
+	Example: `  gh-slack api get conversations.list -f types=public_channel,private_channel
+  gh-slack api post chat.postMessage -b '{"channel":"123","blocks":[...]}`,
+}
+
+var fields []string
+var body string
+
+func init() {
+	apiCmd.Flags().StringSliceVarP(&fields, "field", "f", nil, "Fields to pass to the api call")
+	apiCmd.Flags().StringVarP(&body, "body", "b", "{}", "Body to send as JSON")
+	apiCmd.Flags().StringP("team", "t", "", "Slack team name (required here or in config)")
+	apiCmd.SetHelpTemplate(apiCmdUsage)
+	apiCmd.SetUsageTemplate(apiCmdUsage)
+}
+
+func mapFields(fields []string) (map[string]string, error) {
+	mappedFields := map[string]string{}
+
+	for _, field := range fields {
+		parts := strings.SplitN(field, "=", 2)
+
+		if len(parts) != 2 || parts[1] == "" {
+			return nil, fmt.Errorf("field '%s' is missing a value", field)
+		}
+
+		mappedFields[parts[0]] = parts[1]
+	}
```

Comment:
All those yield the same error:

```
-f types=public_channel,private_channel
-f 'types=public_channel,private_channel'
-f types='public_channel,private_channel'
```

Error:
```
Error: field 'private_channel' is missing a value
```

Reply:
Let's try to follow the `gh api` conventions: https://cli.github.com/manual/gh_api

Would this help?

> To pass nested parameters in the request payload, use "key[subkey]=value" syntax when declaring fields. To pass nested values as arrays, declare multiple fields with the syntax "key[]=value1", "key[]=value2". To pass an empty array, use "key[]" without a value.

Reply:
I solved it with https://github.com/spf13/cobra/issues/661#issuecomment-469223372

Reply:
but that's probably better to do in the `gh` way, I'll have a look at their code

Reply:
They also use `StringArrayVarP`: https://github.com/cli/cli/blob/trunk/pkg/cmd/api/api.go#L231-L232

However they add a bunch of parsing in https://github.com/cli/cli/blob/trunk/pkg/cmd/api/fields.go#L105-L109

I would personally stick to the cobra default for now, with the caveat that if it really adds benefits in the future, we can probably have some inspiration from this file.